RequestService/RequestService.Handlers/PutUpdateJobStatusToAppliedForHandler.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            _repository = repository;
+            _communicationService = communicationService;
+            _groupService = groupService;
+            _jobService = jobService;
+        }
+
+        public async Task<PutUpdateJobStatusToAppliedForResponse> Handle(PutUpdateJobStatusToAppliedForRequest request, CancellationToken cancellationToken)
+        {
+            PutUpdateJobStatusToAppliedForResponse response = new PutUpdateJobStatusToAppliedForResponse()
+            {
+                Outcome = UpdateJobStatusOutcome.Unauthorized
+            };
+
+            bool hasPermission = (request.CreatedByUserID == request.VolunteerUserID || await _jobService.HasPermissionToChangeStatusAsync(request.JobID, request.CreatedByUserID, false, cancellationToken));
+
+            if(!hasPermission)
+            {
+                return response;
+            }
+
+            if (_repository.JobIsInProgressWithSameVolunteerUserId(request.JobID, request.VolunteerUserID) || 
+                (_repository.VolunteerHasAlreadyJobForThisRequestWithThisStatus(request.JobID, request.VolunteerUserID, JobStatuses.AppliedFor)))
+            {
+                response.Outcome = UpdateJobStatusOutcome.AlreadyInThisStatus;
+                return response;
+            }
+            
+            var jobDetails = _repository.GetJobDetails(request.JobID);
+            var volunteerGroups = await _groupService.GetUserGroups(request.VolunteerUserID, cancellationToken);
+            var jobGroups = await _repository.GetGroupsForJobAsync(request.JobID, cancellationToken);
+            int referringGroupId = await _repository.GetReferringGroupIDForJobAsync(request.JobID, cancellationToken);
+
+            if (volunteerGroups == null || jobGroups == null)
+            {
+                throw new Exception("volunteerGroups or jobGroup is null");
+            }
+
+            bool jobGroupContainsVolunteerGroups = jobGroups.Any(volunteerGroups.Groups.Contains);
+
+            if (!jobGroupContainsVolunteerGroups)
+            {
+                response.Outcome = UpdateJobStatusOutcome.BadRequest;
+                return response;
+            }
+
+            var groupMember = await _groupService.GetGroupMember(new GetGroupMemberRequest()
+            {
+                AuthorisingUserId = request.CreatedByUserID,
+                UserId = request.VolunteerUserID,
+                GroupId = referringGroupId
+            });
+
+            var groupActivityCredentials = await _groupService.GetGroupActivityCredentials(new GetGroupActivityCredentialsRequest()
+            {
+                GroupId = referringGroupId,
+                SupportActivityType = new SupportActivityType() { SupportActivity = jobDetails.JobSummary.SupportActivity }
+            });
+
+            bool hasValidCredentials = true;
+
+            foreach (List<int> c in groupActivityCredentials.CredentialSets)
```

Comment:
Not urgent or important, but this could be written as a linq expression (which may end up more or less readable than the for loop...).